aptos-move/aptos-workspace-server/src/services/indexer_api.rs

```diff
@@ -0,0 +1,256 @@
+            "HASURA_GRAPHQL_DEV_MODE=true".to_string(),
+            "HASURA_GRAPHQL_ENABLE_CONSOLE=true".to_string(),
+            // See the docs for the image, this is a magic path inside the
+            // container where they have already bundled in the UI assets.
+            "HASURA_GRAPHQL_CONSOLE_ASSETS_DIR=/srv/console-assets".to_string(),
+            format!("HASURA_GRAPHQL_SERVER_PORT={}", HASURA_DEFAULT_PORT),
+        ]),
+        ..Default::default()
+    };
+
+    let options = CreateContainerOptions {
+        name: format!("aptos-workspace-{}-indexer-api", instance_id),
+        ..Default::default()
+    };
+
+    (options, config)
+}
+
+/// Starts the indexer API service, running in a docker container.
+///
+/// Prerequisites
+/// - Previous task to create the docker network
+///   - Needs to be the same one the postgres container connects to
+/// - Postgres DB (container)
+/// - Indexer processors
+///
+/// The function returns three futures:
+/// - One that resolves to the host port that can be used to access the indexer API service
+///   when it's fully up.
+/// - One that resolves when the container stops  (which it should not under normal operation).
+/// - A cleanup task that stops the container and removes the associated data volume.
+///
+/// As the caller, you should always await the cleanup task when you are ready to shutdown the
+/// service. The cleanup is a "best-effort" operation -- success is not guaranteed
+/// as it relies on external commands that may fail for various reasons.
+pub fn start_indexer_api(
+    instance_id: Uuid,
+    shutdown: CancellationToken,
+    fut_docker_network: impl Future<Output = Result<String, Arc<anyhow::Error>>>
+        + Clone
+        + Send
+        + 'static,
+    fut_postgres: impl Future<Output = Result<u16, Arc<anyhow::Error>>> + Clone + Send + 'static,
+    fut_all_processors_ready: impl Future<Output = Result<(), Arc<anyhow::Error>>>
+        + Clone
+        + Send
+        + 'static,
+) -> (
+    impl Future<Output = Result<u16>>,
+    impl Future<Output = Result<()>>,
+    impl Future<Output = ()>,
+) {
+    let fut_container_clean_up = Arc::new(Mutex::new(None));
+
+    let fut_create_indexer_api = make_shared({
+        let fut_container_clean_up = fut_container_clean_up.clone();
+
+        async move {
+            let (docker_network_name, _postgres_port, _) = try_join!(
+                fut_docker_network,
+                fut_postgres,
+                fut_all_processors_ready
+            )
+            .map_err(anyhow::Error::msg)
+            .context(
+                "failed to start indexer api server: one or more dependencies failed to start",
+            )?;
+
+            println!("Starting indexer API..");
+
+            let (options, config) =
+                create_container_options_and_config(instance_id, docker_network_name);
+            let (fut_container, fut_container_cleanup) =
```

Comment:
Correct me if I'm wrong but I notice you don't try to clear out any leftover containers before creating the new one, e.g. if the clean up failed or the user siginted it or whatever. Should you be doing so here? Or I suppose it's not necessary because you'll use different names and networks each time?

Reply:
> I suppose it's not necessary because you'll use different names and networks each time?

Correct, every `aptos-workspace-server` instance creates its own network, volumes and containers, so there aren't really leftovers (from self) to clean up.

However I do think we should add a separate pass to clean up leftovers globally, in case some previous runs aborted abnormally. Still need to figure out a few more details:
- When do we run it? During localnet start up or as a global pass when you hit the `workspace test` command?
- When do we consider a docker resource removable? More than 2 hours since it was created?
- Do we need some sort of global lock?

Reply:
Hard to say when it's removable safely yeah, maybe just a `workspace prune` command might be better.

Reply:
iiuc, we create a docker resource for each test suite, and after the test finishes we technically dont need it anymore. So we can remove the resource in the "after" hook in workspace (that runs after each test suite is finish)

Reply:
> So we can remove the resource in the "after" hook in workspace (that runs after each test suite is finish)

@0xmaayan Yeah on the TS-side we need to make sure SIGNTERM gets sent to `aptos-workspace-server` binary once it's no longer in use, so it can clean up the resources. I'll double check this part later today.